core/src/main/java/org/frankframework/receivers/Receiver.java

```diff
@@ -1514,86 +1530,129 @@ private Message compactMessage(Message message, PipeLineSession session) throws
 		return msgBuilder.build();
 	}
 
-	@SuppressWarnings("synthetic-access")
-	public synchronized ProcessResultCacheItem cacheProcessResult(String messageId, String errorMessage, Instant receivedDate) {
-		final ProcessResultCacheItem prci = processResultCache.computeIfAbsent(messageId, key -> {
-			log.debug("{} caching first result for messageId [{}]", this::getLogPrefix, ()->messageId);
-			ProcessResultCacheItem item = new ProcessResultCacheItem();
-			item.receiveDate = receivedDate;
+	/**
+	 * Call this method exactly once per every try to process a message to initialize the internally cached receive-count for the messageId.
+	 * If the listener implements {@link IKnowsDeliveryCount} then the receiveCount will be set to the {@link IKnowsDeliveryCount#getDeliveryCount(RawMessageWrapper)}.
+	 *
+	 * @param rawMessageWrapper The raw message for which to set the receiveCount.
+	 */
+	protected synchronized @Nonnull ProcessStatusCacheItem updateMessageReceiveCount(@Nonnull RawMessageWrapper<M> rawMessageWrapper) {
+		String messageId = Objects.requireNonNull(rawMessageWrapper.getId(), () -> "Message must have an ID! No ID for raw message [" + rawMessageWrapper + "]");
+		// We need to know here if a result was previously cached, otherwise we cannot reliably maintain the receiveCount for listeners that don't know the deliveryCount.
+		final ProcessStatusCacheItem prci = processStatusCache.computeIfAbsent(messageId, key -> {
+			log.debug("{} caching first status for messageId [{}]", this::getLogPrefix, ()->messageId);
+			ProcessStatusCacheItem item = new ProcessStatusCacheItem();
+			item.receiveCount = 0;
+			item.receiveDate = Instant.now();
 			return item;
 		});
-		prci.receiveCount++;
-		if (prci.receiveCount > 1) {
-			log.debug("{} increased try count for messageId [{}] to [{}]", this::getLogPrefix, ()->messageId, ()->prci.receiveCount);
+		if (getListener() instanceof IKnowsDeliveryCount<M> knowsDeliveryCount) {
+			prci.receiveCount = knowsDeliveryCount.getDeliveryCount(rawMessageWrapper);
+		} else {
+			prci.receiveCount++;
+		}
+		return prci;
+	}
+
+	@SuppressWarnings("synthetic-access")
+	private synchronized @Nonnull ProcessStatusCacheItem cacheProcessResult(@Nonnull RawMessageWrapper<M> rawMessageWrapper, @Nullable String errorMessage, @Nonnull Instant receivedDate) {
+		final ProcessStatusCacheItem prci = getCachedProcessStatus(rawMessageWrapper);
+		if (prci.receiveCount == 1) {
+			// Set the receiveDate only on first processing of message, to the original receiveDate.
+			prci.receiveDate = receivedDate;
 		}
 		prci.comments = errorMessage;
-		processResultCache.put(messageId, prci);
 		return prci;
 	}
 
-	private synchronized Optional<ProcessResultCacheItem> getCachedProcessResult(String messageId) {
-		return Optional.ofNullable(processResultCache.get(messageId));
+	private synchronized @Nonnull ProcessStatusCacheItem getCachedProcessStatus(@Nonnull RawMessageWrapper<M> rawMessageWrapper) {
```

Comment:
Deze 'cached' status is niet altijd gecached? ik vind de methode `updateMessageReceiveCount` naam lastig, ook omdat hij de `processStatusCache` gebruikt. Kan je in plaats van 2x `computeIfAbsent` een methode maken die een item aanmaakt (en ophaalt)?

Reply:
Dat ophalen gaat wel eens mis doordat de map cache size 0 kan zijn (of anderszins te laag). Ik had eerst maar 1 methode die dit aanmaakte -- en toen voelen allerlei Larva tests om.